zeebe/gateway-rest/src/main/java/io/camunda/zeebe/gateway/rest/deserializer/FilterDeserializer.java

```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Camunda License 1.0. You may not use this file
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.zeebe.gateway.rest.deserializer;
+
+import com.fasterxml.jackson.core.TreeNode;
+import com.fasterxml.jackson.databind.JsonDeserializer;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.io.IOException;
+
+public abstract class FilterDeserializer<T> extends JsonDeserializer<T> {
+
+  protected static final ObjectMapper OM = new ObjectMapper();
```

Comment:
🔧 For later: I think the mapper needs to come from the implementing class because those will sit in the Spring config classes and inject the mapper. Could just be an instance attribute that is passed in a constructor.

Reply:
One thing is important: we need a mapper that doesn't have the same deserializer registered because that would cause a loop.

Reply:
OK, I see. Alternative idea:
* We DO use interfaces for `oneOf`.
* We register the custom deserializers on the `IntegerFilter`.
* If it's a primitive, we return an instance of the `AdvancedIntegerFilter` with `$eq` set.
* Otherwise, we use the one general OM to automatically create an `AdvancedIntegerFilter`.

Assumptions:
* The OM with the registered custom deserializers works fine with implementing classes, not running into a loop because it's registered on the interface.
* The interface usage doesn't run into any other issues (but we handle the deserialization ourselves, so maybe that is fine).

Reply:
Tried two solutions they're both in the commits and they both could work:

1. Define two object mappers. Register deserializers only with one and pass the basic OM as parameter. It works but would need a cleaner solution with the configuration and due to multiple OMs available.
2. Generate interface and implementation with openapi-generator. Register deserializers for the interfaces only and deserialize into the implementation's class. Inject ObjectMapper and use that for deserializing explicit use-cases automatically. It works, changes are also quite miminal.

Decision: 2nd, unless something pops up in CI.

Reply:
Both approaches seem to have some issues still, looking at the CI results, unfortunately.

For the first one, I thinks it's mostly the setup in Spring. We already have a definition for the OM in Spring and we should simply extend that one with the code you defined in the new configuration, I believe.

For the interface approach, I am not 100% sure yet what fails there. Debugging this on the server side would be interesting, I guess, to see where exactly this fails.
Here are the server logs for this: 
[Test results for [IT] Root Integration Tests.zip](https://github.com/user-attachments/files/17480018/Test.results.for.IT.Root.Integration.Tests.zip)

(For future reference, you can access them from the failed [workflow run summary](https://github.com/camunda/camunda/actions/runs/11463393731), scrolling to the bottom and downloading from the **Artifacts**)

Reply:
This is what I can see in the logs there
```
Caused by: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of `io.camunda.zeebe.gateway.protocol.rest.LongFilter` (no Creators, like default constructor, exist): abstract types either need to be mapped to concrete types, have custom deserializer, or contain additional type information
 at [Source: REDACTED (`StreamReadFeature.INCLUDE_SOURCE_IN_LOCATION` disabled); line: 1, column: 33] (through reference chain: io.camunda.zeebe.gateway.protocol.rest.ProcessInstanceSearchQueryRequest["filter"]->io.camunda.zeebe.gateway.protocol.rest.ProcessInstanceFilterRequest["processInstanceKey"])
	at com.fasterxml.jackson.databind.exc.InvalidDefinitionException.from(InvalidDefinitionException.java:67) ~[jackson-databind-2.18.0.jar:2.18.0]
	at com.fasterxml.jackson.databind.DeserializationContext.reportBadDefinition(DeserializationContext.java:1888) ~[jackson-databind-2.18.0.jar:2.18.0]
	at com.fasterxml.jackson.databind.DatabindContext.reportBadDefinition(DatabindContext.java:414) ~[jackson-databind-2.18.0.jar:2.18.0]
	at com.fasterxml.jackson.databind.DeserializationContext.handleMissingInstantiator(DeserializationContext.java:1375) ~[jackson-databind-2.18.0.jar:2.18.0]
	at com.fasterxml.jackson.databind.deser.AbstractDeserializer.deserialize(AbstractDeserializer.java:274) ~[jackson-databind-2.18.0.jar:2.18.0]
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129) ~[jackson-databind-2.18.0.jar:2.18.0]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:399) ~[jackson-databind-2.18.0.jar:2.18.0]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:185) ~[jackson-databind-2.18.0.jar:2.18.0]
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129) ~[jackson-databind-2.18.0.jar:2.18.0]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:399) ~[jackson-databind-2.18.0.jar:2.18.0]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:185) ~[jackson-databind-2.18.0.jar:2.18.0]
	at com.fasterxml.jackson.databind.deser.DefaultDeserializationContext.readRootValue(DefaultDeserializationContext.java:342) ~[jackson-databind-2.18.0.jar:2.18.0]
	at com.fasterxml.jackson.databind.ObjectReader._bindAndClose(ObjectReader.java:2125) ~[jackson-databind-2.18.0.jar:2.18.0]
	at com.fasterxml.jackson.databind.ObjectReader.readValue(ObjectReader.java:1501) ~[jackson-databind-2.18.0.jar:2.18.0]
	at org.springframework.http.converter.json.AbstractJackson2HttpMessageConverter.readJavaType(AbstractJackson2HttpMessageConverter.java:395) ~[spring-web-6.1.13.jar:6.1.13]
```

I think the registration of the deserializers in the setup didn't really work. It looks like they are not picked up, maybe.

Reply:
Interesting, the setup is exactly the same as previously without interfaces and that was green. I suspect it's due to the interface & implementation. It does work fine in `ProcessInstanceQueryControllerTest` though. I'll investigate it a bit.

Reply:
Yeah, I was also confused by that a bit. Let me know if you'd like to pair on this, happy to jump on a call if you need a second pair of eyes.

Reply:
Ohhh, OK 😄

Reply:
For now I took out the `MappingJackson2HttpMessageConverter` from operate. I suspect this will trigger other errors.